### ISSUE TITLE
coqdoc generates useless self-referential links for Context `{…} declarations

### ISSUE THREAD

**Issue by brandenburg** [coq/coq#18468]
#### Description of the problem

`coqdoc` generates useless self-referential links when encountering context declarations of the form:

```coq
Context `{SomeTypeClass ...}.
``` 
Instead of linking to the definition of `SomeTypeClass`, the generated link refers to itself, which is obviously not useful. I tracked this down to a questionable entry in `*.glob` files, but I'm not quite sure how to fix or debug this any further. I would appreciate some pointers. 

#### Coq Version

```
$ coqdoc --version
This is coqdoc version 8.18.0
```

#### Example

Run the following example through `coqdoc`: 

```coq
Class Foo (T : Type) := foo : T -> nat.

Section Named.

  (** Here, coqdoc correctly links [Foo bool] in the line below to the defintion
      above. *)
  Context {FooInstance : Foo bool}.

End Named.


Section AutoNamed.

  (** In the below line, coqdoc links [Foo bool] to the line itself, which is not
      very helpful. (Notably, the link embedded in the comment is correct.) *)
  Context `{Foo bool}.

End AutoNamed.
```

As the comments state, the generated HTML for the declaration in the `AutoNamed` section is pointless:

```html
&nbsp;&nbsp;<span class="id" title="keyword">Context</span> `{
<a id="H:5" class="idref" href="#H:5" name="H:5"><span class="id" title="binder">Foo</span></a>
<a id="H:5" class="idref" href="#H:5" name="H:5"><span class="id" title="binder">bool</span></a>
}.
```
Note how _both_ links specify the anchor name `H:5` and the href `#H:5`. This is clearly not right. 

In contrast, here is what it looks like for the working links in the section `Named`:

```html
&nbsp;&nbsp;<span class="id" title="keyword">Context</span> {
<a id="FooInstance:4" class="idref" href="#FooInstance:4" name="FooInstance:4"><span class="id" title="binder">FooInstance</span></a> :
<a class="idref" href="example.context.html#Foo"><span class="id" title="class">Foo</span></a>
<a class="idref" href="http://coq.inria.fr/distrib/V8.18.0/stdlib//Coq.Init.Datatypes.html#bool"><span class="id" title="inductive">bool</span></a>}.
```

#### Likely Cause

Here's the corresponding `.glob` file:

```
DIGEST 4997b59015d6605b0d8ac530cee5dca3
Fexample.context
ind 7:9 <> Foo
constr 25:27 <> foo
rec 7:9 <> Foo
proj 25:27 <> foo
binder 12:12 <> T:1
R32:35 Coq.Init.Logic <> ::type_scope:x_'->'_x not
R31:31 example.context <> T:1 var
R36:38 Coq.Init.Datatypes <> nat ind
sec 50:54 <> Named
R180:182 example.context <> Foo class
R184:187 Coq.Init.Datatypes <> bool ind
binder 166:176 <> FooInstance:4
R196:200 example.context Named <> sec
sec 213:221 <> AutoNamed
R398:400 example.context <> Foo class
R402:405 Coq.Init.Datatypes <> bool ind
binder 398:405 <> H:5
R414:422 example.context AutoNamed <> sec
```

If I manually remove the line `binder 398:405 <> H:5`, then `coqdoc` generates the correct links. Note how the range `398:405` "shadows" the ranges of the two references in the preceding lines. As there is no identifier in the source code in this location, this entry in the glob file looks highly suspicious to me. 

In contrast, in the lines corresponding to the `Named` section, the range given in the line `binder 166:176 <> FooInstance:4` covers only the actual name and not the two preceding lines.  

Inspired by the diff of #17254, I tried looking through [vernac/](https://github.com/coq/coq/tree/master/vernac) a bit, but I'm not familiar enough with the codebase to make out anything. Is this even the right place to look?

**Comment by brandenburg** [comment]
I now understand that `Dumpglob.dump_binding` is called only in `push_name_evn` in [constrintern.ml:507](https://github.com/coq/coq/blob/7a50f87360009f45580c47b4509596ce96f1d86b/interp/constrintern.ml#L507).

Unfortunately, that file is large and `push_name_env` is called in many places. I'll keep looking, but if someone happens to know the right place to look, any hint would be appreciated.

**Comment by SkySkimmer** [comment]
>If I manually remove the line binder 398:405 <> H:5, then coqdoc generates the correct links. Note how the range 398:405 "shadows" the ranges of the two references in the preceding lines. As there is no identifier in the source code in this location, this entry in the glob file looks highly suspicious to me.

It's true that `` `{foo} `` binds a `H` though.

It's this push_name_env https://github.com/coq/coq/blob/7a50f87360009f45580c47b4509596ce96f1d86b/interp/constrintern.ml#L563
but instead of looking at the implementation we should first decide what should happen.

**Comment by brandenburg** [comment]
Thanks for the pointer! 

> instead of looking at the implementation we should first decide what should happen.

As a user, when clicking on `Foo` in ``` `{Foo bool}``` I expect to be taken to the definition of `Foo`, not the implicit `H`. This is especially true when the definition is on some other file — the typical use case is that I want to find out what that type class actually is and don't know where to look for it. 

I agree that if the `H` is referenced somewhere else, I should be taken to the context declaration. 

Perhaps the solution is to restrict the LOC range emitted in the glob file to the leading "`" symbol, so that it doesn't overlap with refs in the "{…}" part?

**Comment by silene** [comment]
> Perhaps the solution is to restrict the LOC range emitted in the glob file to the leading "`" symbol

Yes, that seems sensible.

**Comment by proux01** [comment]
My understanding is that the generated glob file is fine, correctly containing the `R... Foo class` in both casesn but coqdoc does not generate the link out of it in the second case.

**Comment by brandenburg** [comment]
Yes and no. Yes, the generated glob file does contain the relevant information in both cases, and coqdoc fails to generate the link in the second case. No, I'm not sure the generated glob file is fine, as I'm not sure which component is to blame. Either
-  coqdoc needs to be able to deal with overlapping offset ranges (which it seemingly doesn't currently), or
- the glob file generator must not generate overlapping offset ranges in the second case. 

Intuitively, I'm leaning towards the latter interpretation (overlapping offset ranges don't seem that useful...?), but I don't know enough about the intended design to be sure.

**Comment by proux01** [comment]
According to the documentation of the glob format https://github.com/coq/coq/blob/master/interp/dumpglob.mli#L27 they contain two kind of lines:
* references (lines starting with a capital `R`)
* definitions (other lines)

Definitions are meant to be anchor and I don't see any issue with them (even fully) overlapping.

References are meant to be links and I don't see any overlapping in the glob file above.

That's why I'm inclined to think that the issue comes from coqdoc, that puts links where there isn't in the glob file and ignores others.

**Comment by proux01** [comment]
You may also want to try https://github.com/xavierleroy/coq2html or its extended fork https://github.com/yoshihiro503/coq2html which is much simpler than coqdoc and sometimes works better (it also supports foldable proofs and markdown). We are now using that for Analysis and probably soon for mathcomp.
